File: scripts/reference-solutions/longest-repeating-character-replacement.py
import sys
from collections import defaultdict


def main():
    data = sys.stdin.read().split()
    s = data[0]
    k = int(data[1])
    counts = defaultdict(int)
    left = 0
    best = 0
    max_freq = 0
    for right, c in enumerate(s):
        counts[c] += 1
        max_freq = max(max_freq, counts[c])
        while (right - left + 1) - max_freq > k:
            counts[s[left]] -= 1
            left += 1
        best = max(best, right - left + 1)
    print(best)


main()

File: scripts/reference-solutions/longest-repeating-character-replacement.cpp
#include <algorithm>
#include <array>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <set>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;

int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);

  string s;
  int k;
  cin >> s >> k;

  vector<int> counts(26, 0);
  int left = 0;
  int best = 0;
  int maxFreq = 0;
  for (int right = 0; right < static_cast<int>(s.size()); ++right) {
    int index = s[right] - 'A';
    maxFreq = max(maxFreq, ++counts[index]);
    while ((right - left + 1) - maxFreq > k) {
      --counts[s[left] - 'A'];
      ++left;
    }
    best = max(best, right - left + 1);
  }
  cout << best << '\n';
  return 0;
}
